File: resources/patterns/design/write-behind-cache-pattern/write-behind-cache.ts
/**
 * Write-behind cache with a durable dirty set in Redis (TypeScript).
 *
 * Mirrors write_behind_cache.py: dirty keys are tracked in a Redis set so
 * an application restart does not lose the flush queue. `db` is a
 * placeholder — wire it to your own client (pg, Prisma, Drizzle...).
 *
 * Requires: npm install redis (redis v4+).
 */

import { createClient } from 'redis';

const DIRTY_SET = 'writebehind:dirty';
const FLUSH_SET = 'writebehind:flushing';

type BatchItem = { key: string; value: unknown };
type DbWriter = (batch: BatchItem[]) => Promise<void>;

export class WriteBehindCache {
  private flushTimer: NodeJS.Timeout | null = null;

  constructor(
    private client: ReturnType<typeof createClient>,
    private dbWriter: DbWriter,
    private flushIntervalMs = 5000
  ) {
    this.flushTimer = setInterval(() => {
      this.flush().catch((err) => console.error('flush failed', err));
    }, this.flushIntervalMs);
  }

  async write<T>(key: string, value: T, ttl = 3600): Promise<T> {
    await this.client
      .multi()
      .set(key, JSON.stringify(value), { EX: ttl })
      .sAdd(DIRTY_SET, key)
      .exec();
    return value;
  }

  async flush(): Promise<void> {
    // RENAME is atomic: concurrent writes create a fresh DIRTY_SET.
    try {
      await this.client.rename(DIRTY_SET, FLUSH_SET);
    } catch {
      return; // no dirty keys
    }

    const keys = await this.client.sMembers(FLUSH_SET);
    const batch: BatchItem[] = [];
    for (const key of keys) {
      const raw = await this.client.get(key);
      if (raw !== null) batch.push({ key, value: JSON.parse(raw) });
    }

    try {
      if (batch.length > 0) await this.dbWriter(batch);
      await this.client.del(FLUSH_SET);
    } catch (error) {
      // Merge failed keys back into DIRTY_SET for the next cycle.
      await this.client.sUnionStore(DIRTY_SET, [DIRTY_SET, FLUSH_SET]);
      await this.client.del(FLUSH_SET);
      throw error;
    }
  }

  async dirtyCount(): Promise<number> {
    return this.client.sCard(DIRTY_SET);
  }

  async shutdown(): Promise<void> {
    if (this.flushTimer) clearInterval(this.flushTimer);
    await this.flush();
  }
}

// --- Usage example -------------------------------------------------------

declare const db: { query: (sql: string, values: unknown[]) => Promise<unknown> };

export async function demo(userId: string, name: string, email: string) {
  const redisClient = createClient({ url: 'redis://localhost:6379' });
  await redisClient.connect();

  const cache = new WriteBehindCache(
    redisClient,
    async (batch) => {
      const rows = batch.map((b) => b.value as { id: string; name: string; email: string });
      const values = rows.flatMap((r) => [r.id, r.name, r.email]);
      const placeholders = rows
        .map((_, i) => `($${i * 3 + 1}, $${i * 3 + 2}, $${i * 3 + 3})`)
        .join(', ');
      await db.query(
        `INSERT INTO users (id, name, email) VALUES ${placeholders}
         ON CONFLICT (id) DO UPDATE SET name = EXCLUDED.name, email = EXCLUDED.email`,
        values
      );
    },
    3000
  );

  await cache.write(`user:${userId}`, { id: userId, name, email });
  await cache.shutdown();
  await redisClient.quit();
}

File: resources/patterns/design/write-behind-cache-pattern/write_behind_cache.py
"""Write-behind cache with a durable dirty set in Redis.

The dirty set lives in Redis itself (not in process memory), so an
application restart does not lose the flush queue. A background thread
batches pending keys to the database every `flush_interval` seconds.

Requires: pip install redis (use fakeredis for unit tests).
"""

import json
import threading
import time

import redis


class WriteBehindCache:
    """Cache-aside write path + asynchronous batched database persistence."""

    DIRTY_SET = "writebehind:dirty"
    FLUSH_SET = "writebehind:flushing"

    def __init__(self, redis_client: redis.Redis, db_writer, flush_interval: float = 5.0):
        self.redis = redis_client
        self.db_writer = db_writer
        self.flush_interval = flush_interval
        self._running = True
        self._flush_thread = threading.Thread(target=self._flush_loop, daemon=True)
        self._flush_thread.start()

    def write(self, key: str, value: dict, ttl: int = 3600) -> dict:
        # Pipeline the value write and the dirty-key mark into one round trip.
        pipe = self.redis.pipeline()
        pipe.set(key, json.dumps(value), ex=ttl)
        pipe.sadd(self.DIRTY_SET, key)
        pipe.execute()
        return value

    def _flush_loop(self):
        while self._running:
            time.sleep(self.flush_interval)
            self._flush()

    def _flush(self):
        try:
            # Atomic take of the current dirty set; writes during the flush
            # land in a fresh DIRTY_SET and are picked up next cycle.
            self.redis.rename(self.DIRTY_SET, self.FLUSH_SET)
        except redis.ResponseError:
            return  # no dirty keys

        keys = self.redis.smembers(self.FLUSH_SET)
        batch = {}
        for key in keys:
            raw = self.redis.get(key)
            if raw is not None:
                batch[key] = json.loads(raw)

        try:
            if batch:
                self.db_writer(list(batch.values()))
        except Exception:
            # Merge failed keys back so they are retried next cycle.
            self.redis.sunionstore(self.DIRTY_SET, self.DIRTY_SET, self.FLUSH_SET)
            self.redis.delete(self.FLUSH_SET)
            raise

        self.redis.delete(self.FLUSH_SET)

    def flush_now(self):
        self._flush()

    def dirty_count(self) -> int:
        """Pending keys — alert when this grows without bound."""
        return self.redis.scard(self.DIRTY_SET)

    def shutdown(self):
        self._running = False
        self._flush_thread.join(timeout=10)
        self._flush()


def batch_upsert_users(db, users: list[dict]):
    """One multi-row upsert instead of N single-row updates (psycopg-style)."""
    if not users:
        return
    db.executemany(
        """
        INSERT INTO users (id, name, email)
        VALUES (%s, %s, %s)
        ON CONFLICT (id) DO UPDATE
        SET name = EXCLUDED.name, email = EXCLUDED.email
        """,
        [(u["id"], u["name"], u["email"]) for u in users],
    )


if __name__ == "__main__":
    # Demo against fakeredis if installed, else a real server.
    try:
        import fakeredis

        client = fakeredis.FakeStrictRedis(decode_responses=True)
    except ImportError:
        client = redis.Redis(host="localhost", port=6379, decode_responses=True)

    flushed = []
    cache = WriteBehindCache(client, db_writer=flushed.append, flush_interval=0.5)

    cache.write("user:1", {"id": "1", "name": "Ada", "email": "ada@example.com"})
    cache.write("user:2", {"id": "2", "name": "Grace", "email": "grace@example.com"})
    time.sleep(1.0)
    cache.shutdown()

    print(f"Flushed {len(flushed)} batch(es): {flushed}")
